cmd/snapshot_export.go

```diff
@@ -17,38 +28,172 @@ func (o *snapshotExportOutput) toJSON()  { outputJSON(o) }
 func (o *snapshotExportOutput) toText()  { outputText(o) }
 func (o *snapshotExportOutput) toTable() { outputTable(o) }
 
-func init() {
-	snapshotCmd.AddCommand(&cobra.Command{
-		Use:   "export <snapshot ID>",
-		Short: "export snapshot",
-		Long: fmt.Sprintf(`This command exports a volume snapshot.
+var snapshotExportCmd = &cobra.Command{
+	Use:   "export <snapshot ID>",
+	Short: "export snapshot",
+	Long: fmt.Sprintf(`This command exports a volume snapshot.
 
 Supported output template annotations: %s`,
-			strings.Join(outputterTemplateAnnotations(&snapshotExportOutput{}), ", ")),
-		RunE: func(cmd *cobra.Command, args []string) error {
-			if len(args) != 1 {
-				return cmd.Usage()
-			}
-			return output(exportSnapshot(args[0]))
-		},
-	})
+		strings.Join(outputterTemplateAnnotations(&snapshotExportOutput{}), ", ")),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		if len(args) != 1 {
+			return cmd.Usage()
+		}
+
+		filePath, err := cmd.Flags().GetString("download")
+		if err != nil {
+			return err
+		}
+
+		snapshot, err := exportSnapshot(args[0])
+		if err != nil {
+			return err
+		}
+
+		if !cmd.Flags().Changed("download") {
+			return output(&snapshotExportOutput{
+				URL:      snapshot.PresignedURL,
+				Checksum: snapshot.MD5sum,
+			}, nil)
+		}
+
+		filePath, err = downloadExportedSnapshot(filePath, args[0], snapshot.PresignedURL)
+		if err != nil {
+			return err
+		}
+
+		if err = checkExportedSnapshot(filePath, snapshot.MD5sum); err != nil {
+			return err
+		}
+
+		return nil
+	},
 }
 
-func exportSnapshot(snapshotID string) (outputter, error) {
+func exportSnapshot(snapshotID string) (*egoscale.ExportSnapshotResponse, error) {
 	id, err := egoscale.ParseUUID(snapshotID)
 	if err != nil {
 		return nil, err
 	}
 
-	res, err := asyncRequest(&egoscale.ExportSnapshot{ID: id}, fmt.Sprintf("exporting snapshot %q", id))
+	res, err := asyncRequest(&egoscale.ExportSnapshot{ID: id}, fmt.Sprintf("Exporting snapshot %q", id))
 	if err != nil {
 		return nil, err
 	}
 
-	snapshot := res.(*egoscale.ExportSnapshotResponse)
+	return res.(*egoscale.ExportSnapshotResponse), nil
+}
+
+func downloadExportedSnapshot(filePath, snapshotID, url string) (string, error) {
+	filePath = filepath.ToSlash(filePath)
+
+	if filePath[len(filePath)-1:] == "/" {
+		filePath = filePath + snapshotID
+	}
```

Comment:
If I understand correctly what you do here (i.e. checking if the specified path is a directory, and if so append the snapshot ID to it as filename), it's a rather fragile way to do so: a path can point a directory without being terminated with a trailing slash:

```console
$ mkdir blah
$ stat -c %F blah
directory
$ stat -c %F .
directory
```

Instead, you should call [`os.Stat(filePath)`](https://pkg.go.dev/os?tab=doc#Stat), and if it doesn't return an error – which you'll check using [`os.IsNotExist(err)`](https://pkg.go.dev/os?tab=doc#IsExist) –  check what the returned os.FileInfo's `IsDir()` method tells you instead.

Reply:
Besides, you should return an error telling the user that directory paths are not allowed instead of silently appending the snapshot ID to the directory path.